plugin-server/src/cdp/hog-watcher/hog-watcher.ts

```diff
@@ -0,0 +1,472 @@
+        })
+
+        await this.flushActiveObservations()
+    }
+
+    public getFunctionState(id: HogFunctionType['id']): HogWatcherState {
+        return this.states[id] ?? HogWatcherState.healthy
+    }
+
+    private async checkIsLeader() {
+        const leaderId = await runRedis(this.hub.redisPool, 'getLeader', async (client) => {
+            // Set the leader to this instance if it is not set and add an expiry to it of twice our observation period
+            const pipeline = client.pipeline()
+
+            // TODO: This can definitely be done in a single command - just need to make sure the ttl is always extended if the ID is the same
+
+            // @ts-expect-error - IORedis types don't allow for NX and EX in the same command
+            pipeline.set(`${BASE_REDIS_KEY}/leader`, this.instanceId, 'NX', 'EX', (OBSERVATION_PERIOD * 3) / 1000)
+            pipeline.get(`${BASE_REDIS_KEY}/leader`)
+            const [_, res] = await pipeline.exec()
+
+            // NOTE: IORedis types don't allow for NX and GET in the same command so we have to cast it to any
+            return res[1] as string
+        })
+
+        this.isLeader = leaderId === this.instanceId
+
+        if (this.isLeader) {
+            status.info('👀', '[HogWatcher] I am the leader')
+        }
+    }
+
+    public async sync() {
+        // TODO: Implement this
+        // 1. Expire all old observations and states
+        // 2. Flush any active observations that we need to do redis (and pubsub)
+
+        await this.checkIsLeader()
+        await this.flushActiveObservations()
+
+        if (this.isLeader) {
+            await this.syncState()
+        } else {
+            this.globalState = undefined
+        }
+    }
+
+    private async flushActiveObservations() {
+        const changes: EmittedHogWatcherObservations = {
+            instanceId: this.instanceId,
+            observations: [],
+        }
+
+        const period = periodTimestamp()
+
+        Object.entries(this.currentObservations.observations).forEach(([id, observation]) => {
+            if (observation.timestamp !== period) {
+                changes.observations.push({ id, observation })
+                delete this.currentObservations.observations[id]
+            }
+        })
+
+        if (!changes.observations.length) {
+            return
+        }
+
+        // Write all the info to redis
+        await runRedis(this.hub.redisPool, 'syncWithRedis', async (client) => {
+            const pipeline = client.pipeline()
+
+            changes.observations.forEach(({ id, observation }) => {
+                // We key the observations by observerId and timestamp with a ttl of the max period we want to keep the data for
+                const subKey = `observation:${id}:${this.instanceId}:${observation.timestamp}`
+                pipeline.hset(`${BASE_REDIS_KEY}/state`, subKey, JSON.stringify(observation))
+            })
+
+            return pipeline.exec()
```

Comment:
Where do we set the TTL for these keys... or is that the `TODO: Implement this` part? Isn't the easy solution to just use normal TTLs if this moves from `hset` to `set`?

Reply:
No the todo is just noise I removed in the follow up PR.

The issue is redis 6 (what we use) doesn't support ttl-ing hash fields, only the whole hash.
We could do this as a separate set but for now it just felt easier to have it all in one hash so we can load the whole thing in one go and clean up after.

In practice it functions the same so lets try it and see if it is cleaning up and then we cna move it out after if it still makes sense